state/modelgeneration.go

```diff
@@ -221,6 +221,16 @@ func assignGenerationUnitTxnOps(id, appName, unitName string) []txn.Op {
 // MakeCurrent marks the generation as completed, if it is active and
 // meets autocomplete criteria, so it becomes the "current" generation.
 func (g *Generation) MakeCurrent() error {
+	err := g.complete(false)
+	return errors.Trace(err)
+}
+
+func (g *Generation) Cancel() error {
+	err := g.complete(true)
+	return errors.Trace(err)
+}
+
+func (g *Generation) complete(allowEmpty bool) error {
```

Comment:
Doing it this way, we won't know if the generation was completed or canceled.  Does it matter?

Reply:
It doesn't matter per conversations.  Though changes once we handle history are possible.

Reply:
added a TODO